core/src/main/java/jenkins/util/VirtualFile.java

```diff
@@ -208,6 +314,43 @@
         return callable.call();
     }
 
+    /**
+     * Optionally produces a variant of this handle which may be safely passed over a Remoting {@link Channel}.
+     * This would allow remote nodes such as agents to make calls such as {@link #open}
+     * and be assured of the most efficient possible access.
+     * Otherwise, all calls must be made on the node originally producing this object,
+     * and the caller must arrange for transport of the result.
+     * <p>Note that the result of {@link #forFilePath} does <em>not</em> implement this method,
+     * since a {@link FilePath} may only be transferred over the channel on which it was created.
+     * It cannot, for example, be used to represent a workspace file from one agent on another agent.
+     * @return this object or a variant which may be passed over a {@link Channel}, or null if there is no such support
+     * @since FIXME
+     * @see #toExternalURL
+     */
+    public @CheckForNull VirtualFile asRemotable() {
+        return null;
+    }
+
+    /**
+     * Optionally obtains a URL which may be used to retrieve file contents from any process on any node.
+     * For example, given cloud storage this might produce a permalink to the file.
+     * <p>This is only meaningful for {@link #isFile}:
+     * no ZIP etc. archiving protocol is defined to allow bulk access to directory trees.
+     * <p>Any necessary authentication must be encoded somehow into the URL itself;
+     * do not include any tokens or other authentication which might allow access to unrelated files
+     * (for example {@link ArtifactManager} builds from a different job).
+     * <p>Generally this will be harder to implement than {@link #asRemotable},
+     * which would have the opportunity to perform arbitrary preparation for {@link #open}
+     * such as negotiating session authentication.
+     * @return an externally usable URL like {@code https://gist.githubusercontent.com/ACCT/GISTID/raw/COMMITHASH/FILE}, or null if there is no such support
```

Comment:
Just the most familiar download URL example I could come up with—certainly not a choice for artifact storage!

Reply:
Whoever implements this as a weekend project will get 🍻 from me.

Reply:
Adding some 🍺 to the trophy